src/trans/gpu/internal/dealloc_resol_mod.F90

```diff
@@ -68,7 +78,27 @@ SUBROUTINE DEALLOC_RESOL(KRESOL)
 
   CALL SET_RESOL(KRESOL)
 
-  !TPM_FLT
+#ifdef ACCGPU
+!$ACC EXIT DATA DELETE(R) ASYNC(1)
+!$ACC EXIT DATA DELETE(FG,FG%ZAA0,FG%ZAS0) IF(ALLOCATED(FG%ZAA0)) ASYNC(1)
+!$ACC EXIT DATA DELETE(FG,FG%ZAA,FG%ZAS,FG%ZEPSNM) ASYNC(1)
+!$ACC EXIT DATA DELETE(F,F%RLAPIN,F%RACTHE,F%RW) ASYNC(1)
+!$ACC EXIT DATA DELETE(D,D%MYMS,D%NPNTGTB0,D%NPNTGTB1,D%NSTAGT0B,D%NSTAGT1B,D%NSTAGTF,D%NPROCM)&
+!$ACC&          DELETE(D%NPTRLS,D%MSTABF,D%NASM0,D%OFFSETS_GEMM1,D%OFFSETS_GEMM2) ASYNC(1)
+!$ACC EXIT DATA DELETE(G,G%NDGLU,G%NMEN,G%NLOEN) ASYNC(1)
+!$ACC WAIT(1)
+#endif
+#ifdef OMPGPU
+#endif
+
+  ! Empty all fields (none of them has pointers; allocatable arrays implicitly deallocate)
+  D = D_
+  F = F_
+  FG = FG_
+  R = R_
+  G = G_
```

Comment:
Would it not be simpler to nullify these pointers? I.e. `R => NULL()`?

Reply:
This wouldn't be that easy. Yes, you can do that, but you have to go through the arguments of D and deallocate the allocatable arrays, because they are pointing to the DISTR_RESOL and friends class, which is an allocatable of many such D. When you just nullify the pointers here, it will nullify this pointer, but not the content.
I think, this way is easier, especially if we think that any of the classes might get new allocatables, it is one place less to modify.

Reply:
So what happens to e.g. `D%MYMS` here? In the CPU version, there's an explicit `DEALLOCATE(D%MYMS)`, for example. Do we not need to keep these deallocations also in the GPU version so that host-side versions are correctly destroyed?

Also I noticed `D%MYMS` is referenced below here on line 107 -> is that safe?

Reply:
You are right, we have to do it after the if condition.  But yes, the deallocate is not needed, and my impression is that we also want to avoid it if possible. 
https://github.com/ecmwf-ifs/ectrans/blob/develop/src/trans/common/internal/tpm_distr.F90
All members distr_type are allocatable, that means they deallocate automatically.

Reply:
Okay cool, I think I get it now. Equality is a bit weird with Fortran pointers. So as I understand it, `D = D_` is saying "set whatever `D` is pointing to to `D_`. In other words it does `DISTR_RESOL(NCUR_RESOL) = D_`. The old value of `DISTR_RESOL(NCUR_RESOL)` contained allocated members. But because those are now out of scope, and are `ALLOCATABLE`, they are automatically deleted.

Reply:
`= /= =>` 😜